python: Add capture groups for builtin types, builtin attribute decorators, class inheritance, function arguments and definition keywords

elisiariocouto opened the pull request:
Add capture groups for builtin types, builtin attribute decorators, class inheritance, function arguments and definition keywords.

Related to #14892 

Release Notes:

- Improved syntax highlight for Python: new capture groups for `@function.arguments`, `@function.kwargs`, `@type.class.inheritance`, `@keyword.definition`, `@attribute.builtin` and `@type.builtin`.

elisiariocouto committed:
feat(languages/python): Add queries for builtin types, decorators, class inheritance, function arguments and definition keywords.

elisiariocouto commented:
@cla-bot check

osiewicz commented:
Could you post a before and after comparison?

osiewicz commented:
/cc @danilo-leal @iamnbutler ; do we have a precedent for language-specific highlights?

mikayla-maki commented:
> do we have a precedent for language-specific highlights?

I don't think we do, but I think this is a feature that could be raised to be general across languages that support similar syntax, e.g. Swift's argument labels or C#'s named parameters

elisiariocouto commented:
Just rebased this and changed some capture groups to be similar to Neovim:
- `@builtin.decorators` changed to `@attribute.builtin`
- `@builtin.types` changed to `@type.builtin`

I was thinking of changing the `@keyword.definition` to `@keyword.function` but I think `class` and `lambda` are also definition keywords that aren't applied to functions.

I noticed that Zed's capture groups documentation could be enhanced. It would be great if we could include more information about capture groups or language-specific ones. I read somewhere that there's already some work in progress to expand the capture group list. Do you happen to have a link to that?

osiewicz commented on crates/languages/src/python/highlights.scm:
```diff
@@ -221,3 +243,33 @@
+  (type
+    (binary_operator
+      left: (identifier) @type.builtin))
+  (#match? @type.builtin "^(bool|bytearray|bytes|complex|dict|float|frozenset|int|list|memoryview|object|range|set|slice|str|tuple)$")
```
```suggestion
(#any-of? @type.builtin "bool" "bytearray" "bytes" "complex" "dict" "float" "frozenset" "int" "list" "memoryview" "object" "range" "set" "slice" "str" "tuple")
```
This isn't going to use a regex and is probably a bit cleaner.

elisiariocouto replied:
Applied your suggestion

osiewicz commented on crates/languages/src/python/highlights.scm:
```diff
@@ -221,3 +243,33 @@
+] @keyword.definition
+
+((identifier) @attribute.builtin
+  (#match? @attribute.builtin "^(classmethod|staticmethod|property)$"))
```
```suggestion
  (#any-of? @attribute.builtin "classmethod" "staticmethod" "property")
```

osiewicz reviewed (approved):
Thanks! I don't have a link to an expanded capture group docs (I'm not really in the loop on that); I'd say we should be good to merge this and if anything we can refine capture groups names down the line. Does that sound good to you?

elisiariocouto commented:
Sure thing :)

elisiariocouto commented:
Sorry, pushed a version of the highlights file with a query I'm developing in `tree-sitter-python` to support string prefixes. Force-pushed once again, should be working now.

elisiariocouto commented:
Just a gentle reminder @osiewicz

osiewicz commented:
Whoops, sorry!
